Addresses the problem of an HA instance being partitioned away from the cluster

digitalstain opened the pull request:
This is a set of changes that make explicit the behaviour of instances or groups of instances that are no longer in contact with a majority of the HA cluster. The various commits introduce the ability to detect that state, act on it and revert to read only mode and help with the handling of some edge cases.

changelog: [2.3, HA] Partitioned instances revert to PENDING state while allowing read only operations

digitalstain committed:
Addresses the problem of an HA instance being partitioned away from the cluster

Any instance will now acknowledge the case of suspecting every other instance,
 effectively interpreting it as a case of lost connectivity. The reaction
 now will be to mark all other cluster members as failed, which leads to
 this instance detecting the loss of quorum and therefore becoming unavailable.
 Restoration of connectivity to sufficient instances will make the instance
 available again.
Given the above behaviour, and to handle the case where the current
 master is the instace partitioned away, when a slave detects that the
 master has become unavailable it will set itself to PENDING state, waiting
 the election of a new one.
Finally, some tests had to be changed because previously the assumption
 was that a partitioned instance would maintain its status, which is
 no longer true.

Backported from 3.0. Forward merge with caution.

Conflicts:
	enterprise/cluster/src/main/java/org/neo4j/cluster/protocol/atomicbroadcast/multipaxos/context/HeartbeatContextImpl.java
	enterprise/cluster/src/test/java/org/neo4j/cluster/protocol/heartbeat/HeartbeatStateTest.java
	enterprise/ha/src/test/java/org/neo4j/kernel/ha/cluster/HighAvailabilityMemberStateMachineTest.java

digitalstain committed:
HA cluster members will now become read only on disconnect from cluster

As a means to ensure safety, instances that get detached from a majority
 of the cluster will revert to PENDING state. That implicitly raises the
 barrier disallowing transactions to start, exluding that way the
 possibility of doing read only operations, which are still possible
 in principle, albeit forfeiting up to date data.
This commit makes it so that cluster members which can no longer
 participate in quorum will revert to PENDING state, without
 losing the ability to perform reads. Writes are still of course
 disallowed.

As a side note, a test that tries to verify the behaviour of instances
 when connectivity to quorum is lost abruptly is included but @Ignored.
 This is because while the behaviour tested is part of this patch, the
 fact that it fails is a bug in previous changes. That bug will be
 solved in a different commit and the test will be enabled.

digitalstain committed:
Solves the problem of detecting failures when the cluster suffers partitions

HeartbeatContextImpl will now consult its suspicions before making a
 decision whether a suspected instance is actually failed or not. Previously,
 only failed instances would be ignored when determining quorum of
 suspicions. Now, suspected instances will not be counted towards the
 required suspicion cap, effectively allowing minority partitions to
 mark all disconnected instances as failed and eventually have HA
 move to PENDING state.

MishaDemianenko commented on enterprise/ha/src/test/java/org/neo4j/kernel/ha/ClusterPartitionIT.java:
```diff
@@ -144,6 +142,7 @@ public void losingQuorumIncrementallyShouldMakeAllInstancesPendingAndReadOnly()
                 .withProvider( ClusterManager.clusterOfSize( clusterSize ) )
                 .withSharedConfig( stringMap(
                         ClusterSettings.heartbeat_interval.name(), "1",
+                        ClusterSettings.heartbeat_timeout.name(), "3",
```
maybe it's too low for aws machines

digitalstain replied:
Unfortunately that may be true. I'll amend.

digitalstain committed:
Adds sender instanceid when logging received suspicion message

digitalstain committed:
Illegal transitions of HA state machine are now properly handled

When a message is received which is not expected given the current state,
 instead of throwing an exception, the cluster member transitions to
 PENDING state and asks for an election. This results in what is
 practically a reset of the instance to a proper state by forcing
 a cluster wide serialization of transition events.

MishaDemianenko commented on enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/HighAvailabilityMemberState.java:
```diff
@@ -346,4 +378,23 @@ public abstract HighAvailabilityMemberState masterIsAvailable( HighAvailabilityM
 
     public abstract boolean isAccessAllowed();
+
+    private boolean transitionWasLegal = true;
```
please move those fields to be before any methods, just after SLAVE enum value declaration finish

MishaDemianenko commented on enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/HighAvailabilityMemberStateMachine.java:
```diff
@@ -323,5 +357,32 @@ private long getTotalCount()
             return Iterables.count( members.getMembers() );
         }
+
+        private boolean checkNewStateAndTransition( HighAvailabilityMemberState newState )
```
i think we need to spilt it into 2 separate methods and keep assignment of new state outside of this method

MishaDemianenko replied:
@digitalstain still previous comment, wdyt?

MishaDemianenko commented on enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/HighAvailabilityMemberStateMachine.java:
```diff
@@ -348,5 +357,27 @@ private long getTotalCount()
+                context.setElectedMasterId( null );
+                context.setAvailableHaMasterId( null );
+                changeStateToPending();
+                election.performRoleElections();
```
looks like `return false` is gone? hmmm

MishaDemianenko commented:
@digitalstain another round of review done, couple of question/comments
